Added Set Cursor support for Wasm

Firefly74940 opened the pull request:
GitHub Issue (If applicable): #
<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
Feature 
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Other... Please describe: -->


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
Nothing happend as it's "NotImplemented" part

## What is the new behavior?
Support of different cursors on WASM


## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information
<!-- Please provide any additional information if necessary -->

Internal Issue (If applicable):
<!-- Link to relevant internal issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

It's a draft, code hasn't been tested yet ( need some downloads to be able to build uno.ui and test it )
for now draft is mostly about having feedback on : 

CoreCursorType : Should it be on it's own file ? 

TS setCursor: i did put it in WindowManager ( as it hold many features ) for now, but is it the right place ?

jeromelaban commented:
Thanks for your contribution! You should be able to test it right inside the Samples app that is in the Uno.UI solution. You can [add a sample](https://github.com/nventive/Uno/blob/master/doc/articles/debugging-uno-ui.md#running-the-samples-applications) that sets the pointer type so you can validate that it works properly.

For the typescript, the WindowManager file is getting large and we'll be refactoring it to split it as it gets too large, but for now, you can leave it there.

As for the duplicate type, you can simple move the file from the generated folder to its proper place in the rest of the folder structure.

carldebilly commented on src/Uno.UI.Wasm/ts/WindowManager.ts:
```diff
@@ -1340,6 +1340,14 @@
 		}
+
+		public setCursor( cssCursor: string) {
+			const unoBody = document.getElementById("uno-body");
```
`uno-body`: this element name is supplied in the constructor as `containerElementId`.

Firefly74940 replied:
i simply copied setupSplashScreen method, i will update it ! 

should i also update https://github.com/nventive/Uno/blob/0a54298294d493c1c822f790e54c847523fc04b0/src/Uno.UI.Wasm/ts/WindowManager.ts#L156 in the same commit ?

carldebilly replied:
Yes please ;-)

Firefly74940 replied:
well, turn out as it's a static function, it can't access it, but i guess when the WindowManager code will be refactored, it might move to a static const to solve that particular issue

carldebilly commented on src/Uno.UI.Wasm/ts/WindowManager.ts:
```diff
@@ -1340,6 +1340,14 @@
+			const unoBody = document.getElementById("uno-body");
+
+			if (unoBody) {
+				unoBody.style.cursor = cssCursor;
```
The biggest problem I see with this approach is it won't work for controls deriving from `ButtonBase` because of this css: https://github.com/nventive/Uno/blob/master/src/Uno.UI.Wasm/WasmCSS/Uno.UI.css#L46

The goal was to set the pointer as "pointer" for buttons to improve usuability of such controls in wasm. So changing it globally on the _window_ won't affect buttons and other controls deriving from `ButtonBase`.

Maybe we should also add a property to the root element to temporary deactivate the custom css on buttons while the pointer is _customized_.

Firefly74940 replied:
Using PointerExited / PointerEntered in button code would work ( but then you have to handle it cleverly.. ) [UWP toolkit](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/master/Microsoft.Toolkit.Uwp.UI/Extensions/Mouse/Mouse.cs) provide a way to solve it. 

For my use case, having the property to work everywhere except on button / hyperlink would be ok 

else, spawning a style node and removing it when done would work i guess: 
https://codepen.io/anon/pen/gEbKPB

Firefly74940 commented:
wow, https://github.com/nventive/Uno/pull/629/commits/32f0802e9bdd9cd56bac4831c3020c5d08f0445a contain way too much changes, i will do a clean branch when i've solved my compiles issues

Firefly74940 commented:
Ok,
i've added a sample page
and i've added the technique illustrated here : https://codepen.io/anon/pen/gEbKPB  to override the behaviour pointed by @carldebilly  https://github.com/nventive/Uno/pull/629#discussion_r260398559
i wasn't able to move the CoreCursorType file without getting build error, so i reverted that change

is there anything else i should do before moving the draft to a pull request ?

jeromelaban commented:
This looks good to me, @carldebilly does it address your concern ?

Firefly74940 commented:
if i may up this pr for review @carldebilly 
not that it is urgent or anything, but as @jeromelaban said WindowManager would get a refactor, it would avoid the trouble of merging conflits on that ( or redo pr )

jeromelaban commented:
/azp run

jeromelaban commented:
/azp run

jeromelaban commented:
/azp run

jeromelaban commented:
/azp run

jeromelaban commented:
/azp run

carldebilly commented:
/azp run
